File: spring-data/src/main/java/dev/cerbos/queryplan/springdata/CelDoubleText.java
/*
 * Copyright 2021-2026 Zenauth Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

package dev.cerbos.queryplan.springdata;

import java.math.BigDecimal;
import java.math.MathContext;
import java.math.RoundingMode;
import java.util.regex.Pattern;

/**
 * Inverts CEL's {@code string(double)}, which renders Go's {@code %g} with the shortest digits
 * that round-trip: {@code "-0.6"}, {@code "100000"}, {@code "1e+06"}, {@code "1.5e-05"}.
 *
 * <p>SQL has no portable spelling of that rendering, so {@code string(col) == "text"} is solved
 * in Java for the one double that renders as {@code text}, and the column is compared with it.
 */
final class CelDoubleText {

    private CelDoubleText() {}

    /** Every finite, non-zero rendering: fixed notation, or exponent notation. */
    private static final Pattern RENDERING = Pattern.compile(
            "-?(?:(?:0|[1-9][0-9]*)(?:\\.[0-9]*[1-9])?|[1-9](?:\\.[0-9]*[1-9])?e[+-][0-9]{2,3})");

    /** The spellings CEL gives zero, NaN and the infinities, which a column comparison cannot pin. */
    private static final Pattern UNDECIDABLE = Pattern.compile("-?0|NaN|[+-]Inf");

    /** The outcome of {@link #solve}. */
    sealed interface Solution {
        /** {@code text} is exactly the rendering of {@code value}. */
        record Exactly(double value) implements Solution {}

        /** No double renders as {@code text}. */
        record None() implements Solution {}
    }

    /**
     * The double whose CEL rendering is {@code text}, or {@link Solution.None} when there is
     * none.
     *
     * @throws UnsupportedPlanShapeException when {@code text} is CEL's spelling of zero, NaN or
     *         an infinity, whose SQL comparison cannot tell the cases apart
     */
    static Solution solve(String text) {
        if (UNDECIDABLE.matcher(text).matches()) {
            throw Refusals.unsupported("string() of a number compared with \"" + text + "\": SQL"
                    + " cannot tell the zero, NaN or infinity CEL renders that way from its"
                    + " neighbours");
        }
        if (!RENDERING.matcher(text).matches()) {
            return new Solution.None();
        }
        double value = Double.parseDouble(text);
        if (Double.isInfinite(value)) {
            return new Solution.None();
        }
        String rendered = render(value);
        return text.equals(rendered) ? new Solution.Exactly(value) : new Solution.None();
    }

    /** Go's {@code fmt.Sprintf("%g", value)} for a finite, non-zero double. */
    static String render(double value) {
        BigDecimal exact = new BigDecimal(value);
        BigDecimal digits = shortest(value, exact);
        // Go's decision for the shortest precision: exponent notation when the decimal exponent
        // is below -4 or at least 6, so 100000 stays fixed and 1234567 is 1.234567e+06.
        BigDecimal unscaled = digits.stripTrailingZeros();
        int exponent = unscaled.precision() - unscaled.scale() - 1;
        if (exponent < -4 || exponent >= 6) {
            String mantissa = unscaled.movePointLeft(exponent).abs().toPlainString();
            String sign = exponent < 0 ? "-" : "+";
            String exp = String.valueOf(Math.abs(exponent));
            return (value < 0 ? "-" : "") + mantissa + "e" + sign + (exp.length() < 2 ? "0" + exp : exp);
        }
        return unscaled.toPlainString();
    }

    /**
     * The shortest decimal that parses back to {@code value}. At each length both neighbours of
     * the exact value are tried, since near a power of two the rounding interval is lopsided and
     * the nearer neighbour can fall outside it while the farther one does not.
     */
    private static BigDecimal shortest(double value, BigDecimal exact) {
        for (int p = 1; p <= 17; p++) {
            BigDecimal down = exact.round(new MathContext(p, RoundingMode.FLOOR));
            BigDecimal up = exact.round(new MathContext(p, RoundingMode.CEILING));
            boolean downOk = down.doubleValue() == value;
            boolean upOk = up.doubleValue() == value;
            if (downOk && upOk && down.compareTo(up) != 0) {
                int cmp = exact.subtract(down).compareTo(up.subtract(exact));
                if (cmp == 0) {
                    // Exactly halfway: Go rounds its decimal to even.
                    return down.round(new MathContext(p, RoundingMode.HALF_EVEN))
                            .unscaledValue().testBit(0) ? up : down;
                }
                return cmp < 0 ? down : up;
            }
            if (downOk) {
                return down;
            }
            if (upOk) {
                return up;
            }
        }
        throw Refusals.internal("no 17-digit decimal round-trips " + value);
    }
}

File: exposed/src/main/kotlin/dev/cerbos/queryplan/exposed/CelDoubleText.kt
package dev.cerbos.queryplan.exposed

import java.math.BigDecimal
import java.math.MathContext
import java.math.RoundingMode

/**
 * CEL's `string()` over a double: Go's `strconv.FormatFloat(v, 'g', -1, 64)`, the shortest
 * decimal that round-trips, printed in `%e` form when its exponent is below -4 or at least 6
 * (`1e+06`, `-9.5e+18`) and in plain form otherwise (`2`, `0.5`, `123456`).
 */
internal object CelDoubleText {

    /**
     * The one double CEL prints as [text], or `null` when CEL prints no double that way. A
     * non-finite value is never returned: SQL has no literal for one.
     */
    fun parseCanonical(text: String): Double? {
        val value = text.toDoubleOrNull() ?: return null
        if (!value.isFinite()) return null
        return if (format(value) == text) value else null
    }

    /** [value] as CEL's `string()` prints it. [value] must be finite. */
    fun format(value: Double): String {
        require(value.isFinite()) { "CelDoubleText.format needs a finite value" }
        val negative = value < 0.0 || (value == 0.0 && 1.0 / value < 0.0)
        if (value == 0.0) return if (negative) "-0" else "0"
        val (digits, pointPosition) = shortestDigits(Math.abs(value))
        val exponent = pointPosition - 1
        val body = if (exponent < -4 || exponent >= 6) {
            buildString {
                append(digits[0])
                if (digits.length > 1) append('.').append(digits, 1, digits.length)
                append('e').append(if (exponent < 0) '-' else '+')
                val magnitude = Math.abs(exponent)
                if (magnitude < 10) append('0')
                append(magnitude)
            }
        } else if (pointPosition <= 0) {
            "0." + "0".repeat(-pointPosition) + digits
        } else if (pointPosition >= digits.length) {
            digits + "0".repeat(pointPosition - digits.length)
        } else {
            digits.substring(0, pointPosition) + "." + digits.substring(pointPosition)
        }
        return if (negative) "-$body" else body
    }

    /**
     * The shortest significant digits that round-trip to [value] (positive, finite), and the
     * position of the decimal point relative to them: `value = 0.<digits> × 10^pointPosition`.
     * Rounding the EXACT binary value to each precision in turn finds the nearest decimal of that
     * length, which is the one Go picks among equally short candidates.
     */
    private fun shortestDigits(value: Double): Pair<String, Int> {
        val exact = BigDecimal(value)
        for (precision in 1..17) {
            val rounded = exact.round(MathContext(precision, RoundingMode.HALF_EVEN))
            if (rounded.toDouble() == value) return normalise(rounded)
        }
        return normalise(exact.round(MathContext(17, RoundingMode.HALF_EVEN)))
    }

    private fun normalise(decimal: BigDecimal): Pair<String, Int> {
        val stripped = decimal.stripTrailingZeros()
        val digits = stripped.unscaledValue().toString()
        return digits to (digits.length - stripped.scale())
    }
}
